mchf-eclipse/drivers/audio/audio_agc.c

```diff
@@ -0,0 +1,567 @@
+    agc_wdsp.fast_backmult = 1.0 - expf(-1.0 / (sample_rate * agc_wdsp.tau_fast_backaverage));
+    agc_wdsp.onemfast_backmult = 1.0 - agc_wdsp.fast_backmult;
+
+    agc_wdsp.out_target = agc_wdsp.out_targ * (1.0 - expf(-(float32_t)agc_wdsp.n_tau)) * 0.9999;
+    //  out_target = out_target * (1.0 - expf(-(float32_t)n_tau)) * 0.9999;
+    agc_wdsp.min_volts = agc_wdsp.out_target / (agc_wdsp.var_gain * agc_wdsp.max_gain);
+    agc_wdsp.inv_out_target = 1.0 / agc_wdsp.out_target;
+
+    float32_t tmpA = log10f(agc_wdsp.out_target / (agc_wdsp.max_input * agc_wdsp.var_gain * agc_wdsp.max_gain));
+    if (tmpA == 0.0)
+    {
+        tmpA = 1e-16;
+    }
+    agc_wdsp.slope_constant = (agc_wdsp.out_target * (1.0 - 1.0 / agc_wdsp.var_gain)) / tmpA;
+
+    agc_wdsp.inv_max_input = 1.0 / agc_wdsp.max_input;
+
+    if (agc_wdsp.max_input > agc_wdsp.min_volts)
+    {
+        float32_t convert
+        = powf (10.0, (float32_t)ts.agc_wdsp_conf.hang_thresh / 20.0);
+        float32_t tmpB = (convert - agc_wdsp.min_volts) / (agc_wdsp.max_input - agc_wdsp.min_volts);
+        if(tmpB < 1e-8)
+        {
+            tmpB = 1e-8;
+        }
+        agc_wdsp.hang_thresh = 1.0 + 0.125 * log10f (tmpB);
+    }
+    else
+    {
+        agc_wdsp.hang_thresh = 1.0;
+    }
+
+    float32_t tmpC = powf (10.0, (agc_wdsp.hang_thresh - 1.0) / 0.125);
+    agc_wdsp.hang_level = (agc_wdsp.max_input * tmpC + (agc_wdsp.out_target /
+            (agc_wdsp.var_gain * agc_wdsp.max_gain)) * (1.0 - tmpC)) * 0.637;
+
+    agc_wdsp.hang_backmult = 1.0 - expf(-1.0 / (sample_rate * agc_wdsp.tau_hang_backmult));
+    agc_wdsp.onemhang_backmult = 1.0 - agc_wdsp.hang_backmult;
+
+    agc_wdsp.hang_decay_mult = 1.0 - expf(-1.0 / (sample_rate * agc_wdsp.tau_hang_decay));
+}
+
+
+
+/**
+ *
+ * @param blockSize
+ * @param agcbuffer a pointer to the list of buffers of size blockSize containing the audio data
+ * @param num_channels
+ */
+void AudioDriver_RxAgcWdsp(int16_t blockSize, float32_t (*agcbuffer)[AUDIO_BLOCK_SIZE] )
+{
+#ifdef USE_TWO_CHANNEL_AUDIO
+    const uint8_t dmod_mode = ts.dmod_mode;
+    const bool use_stereo = (dmod_mode == DEMOD_IQ || dmod_mode == DEMOD_SSBSTEREO || (dmod_mode == DEMOD_SAM && ads.sam_sideband == SAM_SIDEBAND_STEREO));
+#endif
+    // Be careful: the original source code has no comments,
+    // all comments added by DD4WH, February 2017: comments could be wrong, misinterpreting or highly misleading!
+    //
+    if (ts.agc_wdsp_conf.mode == 5)  // AGC OFF
+    {
+        for (uint16_t i = 0; i < blockSize; i++)
+        {
+            agcbuffer[0][i] = agcbuffer[0][i] * agc_wdsp.fixed_gain;
+#ifdef USE_TWO_CHANNEL_AUDIO
+            agcbuffer[1][i] = agcbuffer[1][i] * agc_wdsp.fixed_gain;
+#endif
+        }
+        return;
+    }
+
+    for (uint16_t i = 0; i < blockSize; i++)
+    {
+        if (++agc_wdsp.out_index >= agc_wdsp.ring_buffsize)
+        {
+            agc_wdsp.out_index -= agc_wdsp.ring_buffsize;
```

Comment:
There is a member field ring_buffsize. Seems like it is constant, init and never change later. Maybe get rid of it and use AGC_WDSP_RB_SIZE instead?

Reply:
I don't know enough about the WDSP AGC to tell you if there is a reason for changing the size dynamically at some point. You have to ask @DD4WH for more info on this code or checkout the origignal WDSP library.

Reply:
Ok, thanks.